actors/miner/src/expiration_queue.rs

```diff
@@ -776,46 +777,64 @@ impl<'db, BS: Blockstore> ExpirationQueue<'db, BS> {
         sector_size: SectorSize,
         sectors: &[SectorOnChainInfo],
     ) -> anyhow::Result<Vec<SectorExpirationSet>> {
-        let mut declared_expirations = BTreeMap::<ChainEpoch, bool>::new();
+        if sectors.is_empty() {
+            return Ok(Vec::new());
+        }
+
         let mut sectors_by_number = BTreeMap::<u64, &SectorOnChainInfo>::new();
         let mut all_remaining = BTreeSet::<u64>::new();
 
         for sector in sectors {
-            let q_expiration = self.quant.quantize_up(sector.expiration);
-            declared_expirations.insert(q_expiration, true);
             all_remaining.insert(sector.sector_number);
             sectors_by_number.insert(sector.sector_number, sector);
         }
 
-        let mut expiration_groups =
-            Vec::<SectorExpirationSet>::with_capacity(declared_expirations.len());
+        let mut expiration_groups = Vec::<SectorExpirationSet>::with_capacity(sectors.len());
 
-        for (&expiration, _) in declared_expirations.iter() {
-            let es = self.may_get(expiration)?;
+        for sector in sectors {
```

Comment:
Ah, wait. This doesn't make sense. We'll end up iterating over epoch ranges over and over. That's why we iterated over `declaration_expirations` before.

We need to collect a sorted set (btree) of expiration epochs. Then:

```rust
let mut end_epoch = 0;
for &exp in declared_expirations.iter() {
    if exp <= last_end {
        continue
    }
    end_epoch = exp + EPOCHS_IN_DAY;
    self.amt.for_each_while_ranged(Some(exp), ...
}
```

This ensures we:

1. Only scan epoch ranges where interesting sectors may exist.
2. Only scan each range at most once.

Reply:
Good point!  Note that for each sector, we only scan `[sector.expiration, sector.expiration+EPOCHS_IN_DAY]`, which is only 24 hours, so the chance of overlapping is small, even if overlapping, we've excluded it by `all_remaining`.

I'm still considering about it, mainly whether it's worth to introduce a `BTreeSet` to achieve no overlapping at all.

Reply:
> Good point! Note that for each sector, we only scan [sector.expiration, sector.expiration+EPOCHS_IN_DAY], which is only 24 hours, so the chance of overlapping is small, even if overlapping, we've excluded it by all_remaining.

Not really? We'd end up with an overlap as long as any two sectors expire at approximately the same time.

> Ah, I think we can't simply skip if exp <= last_end, because the actual epoch in expiration queue is some quantized result from some deadline, we can only assume it's in range [exp, exp+ EPOCHS_IN_DAY], not more.

Ah.... Yeah, we can't _simply_ skip. We have to:

1. Skip if `new_end <= old_end`.
2. Otherwise, set `exp` to `last_end` if `last_end < new_end`.

Reply:
@Stebalien It seems there're some typos but I get the idea, implemented [here](https://github.com/filecoin-project/builtin-actors/pull/1326/commits/fcc4cc0ed943d5e63f2811c86d0d9dfd1f351b42).

Reply:
Thanks! I've cut a 0.6.2 release for you.